prelude.links

```diff
@@ -1197,13 +1197,21 @@ fun assertEq (actual, expected) server {
 #### APP SERVER ####
 
 sig addSimpleRouteHandler : (String, (String) ~> Page) ~> ()
-fun addSimpleRouteHandler(path,f) { unsafeAddRoute(path, fun(p, _){f(p)}) }
+fun addSimpleRouteHandler(path,f) { addRoute(path,f) }
 
 sig addRoute : (String, (String) ~> Page) ~> ()
-fun addRoute(path, f) { unsafeAddRoute(path, fun(p, _){f(p)}) }
+fun addRoute(path, f) { addLocatedRouteHandler(path, fun(p, _){f(p)}) }
+
+sig defaultErrorPage : (String, String, Location) ~> Page
+fun defaultErrorPage(path, error_string, location) {
+  bodyP(<html><body><h1>Wrong</h1></body></html>)
+}
 
 sig addLocatedRouteHandler : (String, (String, Location) ~> Page) ~> ()
-fun addLocatedRouteHandler(path,f) { unsafeAddRoute(path, fun(p, l){f(p, l)}) }
+fun addLocatedRouteHandler(path,f) { addLocatedRouteWithErrors(path, fun(p, l){f(p, l)}, defaultErrorPage) }
+
```

Comment:
It occurred to me when writing this that the number of ways to add routes was proliferating rapidly, and we might want to avoid a quadratic number of addXRouteWithY functions.  But I didn't have a good alternative approach in mind.